### DISCUSSION TITLE
ESP32 leaves pins connected...

### DISCUSSION THREAD

**Discussion by ericgu** [Makuna/NeoPixelBus#499] [Q&A] [answered]
I'm seeing some weird behavior with settings that are persisting across reboots...

I'm doing a demo setup where I have three different ws2812 strips connected to three different pins. I want to be able to use them one at a time by defining whichever pin I want to use after a reset.

What is happening is weird. If I start with power on, the first string that I connect to works correctly. My code then resets the ESP32, creates a NeoPixelBus on a different pin. It will now send output to the new pin *and* the pin from the previous time.

Then I do the same thing with the third pin, and I get the same output running on 3 pins. 

It seems likely that something isn't getting reset at the ESP32 level. 

Any thoughts?

**Comment by Makuna** [comment, marked as answer]
Which "method" are you using?  How do you create it and destroy it? 

The ESP32 has a hardware pin muxer (means it will assign pin/pins to hardware) and it sounds like these are not being reset by ESP32 when it reboots.  
UPDATE: If this is the case, I can add code to the destructors to turn off the pin mux that is used; but again, you need to make sure to destruct the NeoPixelBus before you reset the device.  If you can't control the reset timing, then you need to add code to your setup/startup to demux the last used pin before using a new pin.

**Reply by ericgu** [reply]
Thanks for the quick response. 

My code looks like this:

        _pStrip = new NeoPixelBus<NeoGrbFeature, Neo800KbpsMethod>(_pixelCount, _pixelPin);

I hadn't realized the ESP-32 had the weird soft reset behavior, so I went through and added cleanup code that will delete my NeoPixelBus instance. 

No change in behavior.

My guess is that the pin assignment being done in the constructor code isn't being undone in the destructor code. 

This should repro by creating an instance on one pin, deleting it, and then creating it on another pin.

You mention demuxing the last pin in my code. I could certainly do that; can you give me a pointer as to what code I would use to do the demux? I'm currently using the Arduino bindings because reasons.

**Reply by Makuna** [reply]
The destructor currently doesn't call anything yet.

Try this, add this line after you destruct the NeoPixelBus
```
gpio_matrix_out(_pixelPin, 0x100, false, false);
```

**Reply by ericgu** [reply]
Thanks. I'll give it a go when I have a chance and get back to you.

**Reply by ericgu** [reply]
Okay, that indeed did work. 

Thanks a bunch for your help.

**Comment by Makuna** [comment]
Note, I created an issue to track adding the demux of the pin when the NeoPixelBus gets destructed; so all you should have to worry about is destructing the object and it gets handled for you.

**Reply by ericgu** [reply]
Thanks for your help. And thanks for the library.
